hts-sys/build.rs

```diff
@@ -43,115 +90,164 @@ fn main() {
+        }
     }
 
     let use_lzma = env::var("CARGO_FEATURE_LZMA").is_ok();
-    if !use_lzma && want_static {
-        let lzma_patterns = vec!["s/ -llzma//", "/#define HAVE_LIBLZMA/d"];
-        sed_htslib_makefile(&out, &lzma_patterns, "lzma");
-    } else if let Ok(inc) = env::var("DEP_LZMA_INCLUDE").map(PathBuf::from) {
-        cfg.include(inc);
+    if use_lzma {
+        if let Ok(inc) = env::var("DEP_LZMA_INCLUDE").map(PathBuf::from) {
+            cfg.include(inc);
+            config_lines.push("#define HAVE_LIBBZ2 1");
+            config_lines.push("#ifndef __APPLE__");
+            config_lines.push("#define HAVE_LZMA_H 1");
+            config_lines.push("#endif");
+        }
     }
 
     let use_curl = env::var("CARGO_FEATURE_CURL").is_ok();
-    if !use_curl {
-        let curl_patterns = vec!["s/ -lcurl//", "/#define HAVE_LIBCURL/d"];
-        sed_htslib_makefile(&out, &curl_patterns, "curl");
-    } else if let Ok(inc) = env::var("DEP_CURL_INCLUDE").map(PathBuf::from) {
-        cfg.include(inc);
+    if use_curl {
+        if let Ok(inc) = env::var("DEP_CURL_INCLUDE").map(PathBuf::from) {
+            cfg.include(inc);
+            config_lines.push("#define HAVE_LIBCURL 1");
+            cfg.file("htslib/hfile_libcurl.c");
+            println!("cargo:rerun-if-changed=htslib/hfile_libcurl.c");
+
+            #[cfg(target_os="macos")]
+            {
+                // Use builtin MacOS CommonCrypto HMAC
+                config_lines.push("#define HAVE_COMMONCRYPTO 1");
+            }
+
+            #[cfg(not(target_os="macos"))]
+            {
+                if let Ok(inc) = env::var("DEP_OPENSSL_INCLUDE").map(PathBuf::from) {
+                    // Must use hmac from libcrypto in openssl
+                    cfg.include(inc);
+                    config_lines.push("#define HAVE_HMAC 1");
+                } else {
+                    panic!("No OpenSSL dependency -- need OpenSSL includes");
+                }
+            }
+        }
     }
-    
 
-    let tool = cfg.get_compiler();
-    let (cc_path, cflags_env) = (tool.path(), tool.cflags_env());
-    let cc_cflags = cflags_env.to_string_lossy().replace("-O0", "");
-    let host = env::var("HOST").unwrap_or_default();
-    // autoreconf & ./configure (with no args) steps are necessary to include the htslib plugins (hfile_s3.o, hfile_s3_writer.o, etc...)
-    if  // cleanup first
-        // TODO: Have top level "cargo clean" do this instead of in here, see:
-        // https://github.com/rust-lang/cargo/issues/572#issuecomment-632456478
-        !Command::new("make")
-        .current_dir(out.join("htslib"))
-        .arg("clean")
-        .status().unwrap().success()
+    let use_gcs = env::var("CARGO_FEATURE_GCS").is_ok();
+    if use_gcs {
+        config_lines.push("#define ENABLE_GCS 1");
+        cfg.file("htslib/hfile_gcs.c");
+        println!("cargo:rerun-if-changed=htslib/hfile_gcs.c");
+    }
+
+    let use_s3 = env::var("CARGO_FEATURE_S3").is_ok();
```

Comment:
@pmarks @nlhepler To be honest, I would not use environment variables for configuration of features but cargo flags instead (as I did above with `static = []`), see https://github.com/alexcrichton/xz2-rs/pull/63 for reference.

Reply:
IMO it's more cargo-intuitive (as a developer) to use `features=` instead of having to export something pre-compilation.

Reply:
@brainstorm this is how cargo features are passed to the build script, see [here](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts)

Reply:
See e.g. https://github.com/rust-lang/libc/blob/master/build.rs#L5-L19 for an example.

Reply:
@nlhepler I hear you and I just thought about additional reasons to use simple cargo flags **in addition to env vars**. 

For instance, we have to define each one of them again under the cross passthrough if we want to have them available during builds and tests:

https://github.com/rust-bio/rust-htslib/blob/master/Cross.toml#L2

If Cargo flags are used (actually, in addition to env vars like in that `xz2-rs` crate I referred to above?), users could just have: **features=["s3", "gcs"]**, etc on their `Cargo.toml`, without the need to modify `Cross.toml` when and if new features are introduced in (rust-)htslib.

Reply:
@brainstorm that's exactly the mechanism that I'm trying to explain here. Cargo builds the build.rs script, and when it runs it, ensures env::vars CARGO_FEATURE_S3 and CARGO_FEATURE_GCS are defined in the build script's execution environment if you passed `--features=s3,gcs` to `cargo b`. There is no mucking around with e.g. `CARGO_FEATURE_S3=1 cargo b --features=s3`. These env vars are the mechanism by which cargo communicates to the build script.

Reply:
Ahaa, sorry @nlhepler, thanks for the clarification and patience ;)

Reply:
:) My bad, I was doing a poor job of explaining. I think the advantage of the CARGO_FEATURE approach is that it doesn't require the build script to be rebuilt upon feature changes, just re-run.

Reply:
Definitely, I'm sold now ;)